kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java

```diff
@@ -147,27 +111,60 @@ private Supplier<CompletionStage<KeyContext>> makeKeyContext(@NonNull K kekId) {
+                                                 @NonNull Receiver receiver, int attempt) {
+        if (attempt >= 3) {
+            return CompletableFuture.failedFuture(new EncryptionException("failed to encrypt records after " + attempt + " attempts"));
+        }
+        return currentDekContext(encryptionScheme.kekId()).thenCompose(keyContext -> {
             synchronized (keyContext) {
-                var valueCiphertext = maxValuePlaintextSize >= 0 ? bufferPool.acquire(keyContext.encodedSize(maxValuePlaintextSize)) : null;
-                try {
-                    encryptRecords(encryptionScheme, keyContext, records, fieldsHeader, dekHeader, valueCiphertext, receiver);
-                }
-                finally {
-                    if (maxValuePlaintextSize >= 0) {
-                        bufferPool.release(valueCiphertext);
+                // if it's not alive we know a previous encrypt call has replaced the stage in the cache and fall through to retry encrypt
+                if (!keyContext.isDestroyed()) {
+                    if (!keyContext.hasAtLeastRemainingEncryptions(records.size())) {
+                        // replace the key context stage in the cache, then call encrypt again
+                        rotateKeyContext(encryptionScheme, keyContext);
+                    }
+                    else {
+                        return encrypt(encryptionScheme, records, receiver, keyContext);
                     }
                 }
             }
+            return attemptEncrypt(encryptionScheme, records, receiver, attempt + 1);
         });
     }
 
+    @NonNull
+    private CompletableFuture<Void> encrypt(@NonNull EncryptionScheme<K> encryptionScheme, @NonNull List<? extends Record> records,
+                                            @NonNull Receiver receiver, KeyContext keyContext) {
+        var fieldsHeader = createEncryptedFieldsHeader(encryptionScheme.recordFields());
+        var dekHeader = createEdekHeader(keyContext);
+        var maxValuePlaintextSize = encryptionScheme.recordFields().contains(RecordField.RECORD_VALUE)
+                ? records.stream().mapToInt(Record::valueSize).max().orElse(-1)
+                : -1;
+        var sharedValueCiphertextBuffer = maxValuePlaintextSize >= 0 ? bufferPool.acquire(keyContext.encodedSize(maxValuePlaintextSize)) : null;
+        try {
+            encryptRecords(encryptionScheme, keyContext, records, fieldsHeader, dekHeader, sharedValueCiphertextBuffer, receiver);
+        }
+        finally {
+            if (sharedValueCiphertextBuffer != null) {
+                bufferPool.release(sharedValueCiphertextBuffer);
+            }
+        }
+        keyContext.recordEncryptions(records.size());
+        return CompletableFuture.completedFuture(null);
+    }
+
+    // this must only be called while holding the lock on this keycontext
```

Comment:
It looks nearly right but I can't see how to refer to the correct lock. It looks geared towards locking on this, or a field, or a class, or something you can extract from a no-args method call.

Reply:
I think it could be expressed as `@Holding("#2")` from the Checker framework.  I'm not suggesting to adopt Checker as part of this PR.  I think a comment is good enough for now.